src/commands/command_runner.py

```diff
@@ -0,0 +1,187 @@
+
+    return verify(joined), joined
+
+
+OnExitCallback = Optional[Callable[[CommandFinalResult], None]]
+OnLineCallback = Optional[Callable[[CommandLineOutput], None]]
+
+
+class CommandRunner:
+    """ Wrapper class that utilizes subprocess.Popen and allows more control over processes
+    Creates a process and runs it lazily (when .run() is called)
+    Allows a function (on_exit) to be called on process end with the return code and output of the process
+    Allows a function (on_line) to be called on each line output from the command
+    Allows the process to be killed (using .kill()) with all sub-processes in the same group id
+    At any time .is_complete() can be called to check if complete, and .output to get the output if complete
+    Status available for each step in the lifecycle of a command
+    """
+
+    def __init__(self, command_id: int, command: str, on_line: OnLineCallback = None, on_exit: OnExitCallback = None):
+        """ Creates a lazy CommandRunner which doesn't start the command until run is called
+        :param command_id: command id which is part both output types
+        :param command: the command that should be run, it must be a string
+        :param on_line: optional callback function that is called for each line resulted on the output,
+                        this is useful for long running processes when keeping track of output is essential
+        :param on_exit: callback function that is called when the process ends,
+                        it is called with the return code and the output of the process
+                        note however that if on_line is supplied,
+                        then the command string output of this line would be empty
+        """
+
+        if not isinstance(command, str):
+            raise Exception("Command should be string")
+
+        if len(command) <= 0:
+            raise Exception("Command content can't be empty")
+
+        self.__command_id = command_id
+        self.__command: list = shlex.split(command)
+        self.__on_line: OnLineCallback = on_line
+        self.__on_exit: OnExitCallback = on_exit
+
+        self.__process_id: Optional[int] = None
+        self.__process: Optional[subprocess.Popen] = None
+        self.__status: CommandStatus = CommandStatus.NotStarted
+        self.__output: Optional[CommandFinalResult] = None
+
+    @property
+    def command_id(self):
+        return self.__command_id
+
+    @property
+    def command(self):
+        return self.__command
+
+    @property
+    def process_id(self):
+        return self.__process_id
+
+    @property
+    def status(self):
+        return self.__status
+
+    @property
+    def output(self):
+        return self.__output
+
+    @property
+    def is_complete(self) -> bool:
+        """ :return: whether the process completed (both successfully or killed) """
+
+        return self.__status == CommandStatus.Killed or self.__status == CommandStatus.FinishedSuccessfully
```

Comment:
i like these 👍

Reply:
They'll even throw errors if someone from outside the class tries to change them as they are getters only.